### ISSUE TITLE
Race Condition in bcm2835gpio_write

### ISSUE THREAD

**Issue by tschesnok** [derekmulcahy/xvcpi#4]
I was able to get a vitis debug session to fail when moving my mouse. Interrupts should technically not be able to disrupt a jtag session since the host controls the clock. 

It seems that in bcm2835gpio_write there is a race condition where the clock is set first before the data. If an interrupt happens (and it does) then you corrupt data.

I changed code to the following and it seems rock solid now. The idea is that you always change the clock bit LAST after the data clock. What is confusing to me is that it seems inverted.. but since it works I thought I would post and get feedback.

if (!tck)
{
        // clock going high so "set" must go last
        GPIO_CLR = clear;
        GPIO_SET = set;
}
else
{
        // clock going low.. so clear must go last
        GPIO_SET = set;
        GPIO_CLR = clear;
}

**Comment by tschesnok** [comment]
Follow up.. no.. does not help. Seems some days it works fine and other days just having the link open corrupts the FPGA or the debug session. But why is the real questions. Definitely a race condition above.. just not the only bug.

**Comment by nic3-14159** [comment]
I looked at a few JTAG timing diagrams, and it seems like the real issue is the TDO timing. It seems like the target device assets TDO on the falling edge of the clock, and the host should sample TDO on the rising edge. In the original code, TDO was being sampled right after setting the clock low, which limited the max clock rate that could be used without issues. I've changed this in pull request #5, which so far has consistently worked fine for me. As for TMS and TDI, the original code was fine, as they were being changed on the falling edge of the clock, whereas the target samples them on the rising edge, during which the original code just maintains their values.
